src/main/java/io/split/android/client/service/sseclient/sseclient/NewSseClientImpl.java

```diff
@@ -0,0 +1,159 @@
+import io.split.android.client.network.HttpStreamResponse;
+import io.split.android.client.network.URIBuilder;
+import io.split.android.client.service.sseclient.EventStreamParser;
+import io.split.android.client.service.sseclient.SseJwtToken;
+import io.split.android.client.utils.Logger;
+import io.split.android.client.utils.StringHelper;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.lang.reflect.Modifier.PRIVATE;
+
+public class NewSseClientImpl implements NewSseClient {
+
+    private final URI mTargetUrl;
+    private AtomicInteger mReadyState;
+    private final HttpClient mHttpClient;
+    private EventStreamParser mEventStreamParser;
+    private AtomicBoolean isDisconnectCalled;
+    private SseHandler mSseHandler;
+
+    private final StringHelper mStringHelper;
+
+    private BufferedReader mBufferedReader;
+    private HttpStreamRequest mHttpStreamRequest = null;
+
+    final static int CONNECTING = 0;
+    final static int CLOSED = 2;
+    final static int OPEN = 1;
+
+    private static final String CONTENT_TYPE_HEADER = "Content-Type";
+    private static final String CONTENT_TYPE_VALUE_STREAM = "text/event-stream";
+    private static final String PUSH_NOTIFICATION_CHANNELS_PARAM = "channel";
+    private static final String PUSH_NOTIFICATION_TOKEN_PARAM = "accessToken";
+    private static final String PUSH_NOTIFICATION_VERSION_PARAM = "v";
+    private static final String PUSH_NOTIFICATION_VERSION_VALUE = "1.1";
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    public NewSseClientImpl(@NonNull URI uri,
+                            @NonNull HttpClient httpClient,
+                            @NonNull EventStreamParser eventStreamParser,
+                            @NonNull SseHandler sseHandler) {
+        mTargetUrl = checkNotNull(uri);
+        mHttpClient = checkNotNull(httpClient);
+        mEventStreamParser = checkNotNull(eventStreamParser);
+        mSseHandler = checkNotNull(sseHandler);
+        mReadyState = new AtomicInteger(CLOSED);
+        isDisconnectCalled = new AtomicBoolean(false);
+        mStringHelper = new StringHelper();
+        mReadyState.set(CLOSED);
+    }
+
+    public int readyState() {
+        return mReadyState.get();
+    }
+
+    @Override
+    public void disconnect() {
+        if (readyState() != CLOSED) {
+            isDisconnectCalled.set(true);
+            setCloseStatus();
+            if (mHttpStreamRequest != null) {
+                mHttpStreamRequest.close();
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        Logger.d("Shutting down SSE client");
+        setCloseStatus();
+        disconnect();
+    }
+
+    private void setCloseStatus() {
+        mReadyState.set(CLOSED);
+    }
+
+    private void setReadyState(int state) {
+        mReadyState.set(state);
+    }
+
+    @Override
+    public void connect(SseJwtToken token, ConnectionListener connectionListener) {
```

Comment:
Good point doc, but this listener is only a callback to inform the connect caller that the connection succeded so that certain setup is done after that. No reference to the caller is kept.
Sse handler handles all incoming notification as Push specs states.